README
== Description
   win32-clipboard - a Ruby library for interacting with the MS Windows clipboard.

== Installation
   gem install win32-clipboard

== Synopsis
   require 'win32/clipboard'
   include Win32
    
   puts "Data on clipboard: " + Clipboard.data
   puts "Number of available formats: " + Clipboard.num_formats.to_s
    
   puts "Setting data to 'foobar'"
   Clipboard.set_data("foobar")
    
   puts "Clipboard now contains: " + Clipboard.data
    
   puts "Clearing clipboard"
   Clipboard.empty
        
== Class Methods
Clipboard.data(format=nil)
   Returns the data currently in the clipboard.  If +format+ is
   specified, it will attempt to retrieve the data in that format. The
   default is Clipboard::TEXT. See the 'Constants' section for
   the supported formats.
   
   If there is no data in the clipboard, then an empty string is returned.
    
   Note that this method does implicit conversions on formats with regards
   to text data. See the MSDN documentation for more details.
    
Clipboard.empty
   Empty the contents of the clipboard
    
Clipboard.format_available?(format_number)
   Returns true if 'format_number' is currently available on the clipboard,
   false otherwise.

Clipboard.format_name(format_number)
   Returns the corresponding name for the given 'format_number', or nil
   if it does not exist.

Clipboard.formats
   Returns a hash of all the current formats, with the format number as the
   key and the format name as the value for that key.

Clipboard.get_data(format=nil)
   Alias for Clipboard.data.
    
Clipboard.num_formats
    Returns the number of different data formats currently on the clipboard.
    
Clipboard.register_format(format)
   Registers the given 'format' (a String) as a clipboard format, which
   can then be used as a valid clipboard format.
	
   If a registered format with the specified name already exists, a new
   format is not registered and the return value identifies the existing
   format. This enables more than one application to copy and paste data
   using the same registered clipboard format. Note that the format name
   comparison is case-insensitive.

   Registered clipboard formats are identified by values in the range 0xC000
   through 0xFFFF.

Clipboard.set_data(data, format=nil)
   Sets the clipboard contents to the data that you specify.  You may
   optionally specify a clipboard format. The default is Clipboard::TEXT.
    
   See the 'Constants' section for a list of the supported formats.

== Constants
=== Standard Constants
VERSION
   Returns the current version number of this package, as a String.
    
=== Clipboard Formats
Clipboard::OEMTEXT
   Text format containing characters in the OEM character set. Each line ends
   with a carriage return/linefeed (CR-LF) combination. A null character
   signals the end of the data.
    
Clipboard::TEXT
   Text format. Each line ends with a carriage return/linefeed (CR-LF)
   combination. A null character signals the end of the data. Use this format
   for ANSI text.
    
Clipboard::UNICODETEXT
   Unicode text format. Each line ends with a carriage return/linefeed
   (CR-LF) combination. A null character signals the end of the data.
    
== Future Plans
   Add more formatting option contants and related methods.
    
== License
   Artistic 2.0
    
== Copyright
   (C) 2003-2015 Daniel J. Berger
   All Rights Reserved
    
== Authors
   Daniel J. Berger
   Park Heesob


CHANGES
== 0.6.4 -19-Oct-2015
* This gem is now signed, and the gem:create task now signs the gem.
* Gem related tasks in the Rakefile now assume Rubygems 2.x.
* Updated the gemspec to include cert info and remove rubyforge_project info.
* Added a win32-clipboard.rb file for convenience.
* Use require_relative where appropriate.

== 0.6.3 - 7-Feb-2014
* Fixed a potential duplication notification issue in the notify_change method.
  Thanks go to u338steven for the spot and patch.
* Fixed a clipboard chaining issue in the notify_change method when there are
  multiple clipboard instances. Thanks again go to u338steven.

== 0.6.2 - 27-Jan-2014
* Fix and more tests for the internal get_image_data. Thanks go to u338steven
  for the spot and patches.

== 0.6.1 - 20-Jan-2014
* Use 32-bit versions of GetWindowLong and SetWindowLong functions on 32-bit
  versions of Ruby. Thanks go to Orgad Shaneh for the spot.

== 0.6.0 - 12-Jan-2014
* Converted code to use FFI.
* Now requires Ruby 1.9.1 or later and Windows Vista or later.
* The HtmlClipboard class has been added. This lets you get and set data
  in html format.
* Added the get_html_data and set_html_data methods to the Clipboard class.
  These are wrappers for the HtmlClipboard methods.
* You can now set bitmap data properly as well as retrieve it. Thanks go to
  Tadashi Kba for the patch.

== 0.5.2 - 14-Jul-2009
* Replaced some missing include's that I thought were unnecessary. Thanks go
  to Liu Song for the spot.
* Changed the license to Artistic 2.0, and set that explicitly in the gemspec.

== 0.5.1 - 26-Apr-2009
* Modified the Clipboard.register_format method to return the integer value
  of the registered format.
* Removed some unnecessary include's.

== 0.5.0 - 24-Apr-2009
* Added the Clipboard.notify_change method, which notifies you if any changes
  have been made to the clipboard.
* Added support for copying file lists, bitmaps and Windows metafiles.
* Added the Clipboard.clear alias for Clipboard.empty
* The previous documentation for the Clipboard.format_available? method was
  misleading. It has been corrected.
* Added rdoc for modules, classes and constants, and added some inline
  code examples for some methods.
* Gemspec update for handling a deprecated method.

== 0.4.4 - 26-Aug-2008
* Some internal refactoring to ensure that the clipboard is closed on failure
  in a more robust fashion.
* Renamed the test file.
* Eliminated the custom memcpy function, since windows-pr now does the right
  thing. This necessitated bumping the minimum required windows-pr version.

== 0.4.3 - 25-Jul-2007
* Changed ClipboardError to Clipboard::Error
* Added a Rakefile with tasks for testing and installation.
* Removed the install.rb file since installation is now handled by the Rakefile.
* Updates to the README and MANIFEST.

== 0.4.2 - 27-Jan-2007
* Fixes for the clipboard_test.rb example program setup code. Thanks go to
  Harold Hausman for the spot.

== 0.4.1 - 30-Jul-2006
* Now uses the windows-pr package.  No API changes.
* Made Clipboard.get_data a true alias (instead of a facade).

== 0.4.0 - 24-Feb-2006
* Removed the C version entirely and replaced it with a pure Ruby version
  that uses the Win32API package.
* Added a gemspec.
* Added a install.rb file.

== 0.3.1 - 2-Feb-2006
* Unicode fix to handle characters with null bytes.  Thanks go to Brian
  Marick for the spot.

== 0.3.0 - 26-May-2005
* Changed the Clipboard.data= method to Clipboard.set_data because it does,
  in fact, take up to two arguments.  This API change is the main reason for
  the VERSION bump.
* Added Clipboard.get_data method as an alias for Clipboard.data.
* Removed the clipboard.rd file.  The Clipboard.txt file is now rdoc friendly.
* Now Unicode friendly.
* Added more tests, including Unicode test.
* General code cleanup.

== 0.2.1 - 28-Feb-2005
* Moved the 'examples' directory to the toplevel directory.
* Renamed the sample script to 'clipboard_test.rb'.
* Made this document and README rdoc friendly.

== 0.2.0 - 12-Jul-2004
* Made Clipboard a class instead of a module.
* Moved the ClipboardError class under the Win32 module.
* Added the formats() class method.
* Added the register_format class method.
* Added the format_available? class method.
* Added the format_name class method.
* Replaced all instances (1) of the deprecated STR2CSTR() function with the
  StringValuePtr() function.  This means that as of version 0.2.0 this package
  requires Ruby 1.8.0 or later.
* Moved the sample script to doc/examples.

== 0.1.0 - 19-Nov-2003
* Initial release


test/lock.rb
require 'timeout'
require 'tmpdir'

def lock(&block)
  Timeout::timeout(5) do
    open(File.join(Dir.tmpdir, 'ruby-win32-clipboard.lock'), 'w') do |f|
      begin
        f.flock(File::LOCK_EX)
        yield
      ensure
        f.flock(File::LOCK_UN)
      end
    end
  end
rescue Timeout::Error
  raise 'Lock(timeout)'
end



test/notify.rb
require 'win32/clipboard'
require File.join(File.dirname(__FILE__), 'lock')

include Win32

timeout = ARGV[1].to_i

def write
  data = ''
  lock do
    data = Clipboard.data
  end

  begin
    STDOUT.puts data
    STDOUT.flush
  rescue => e
    puts $!
  end
  if data == ARGV[0]
    exit
  end
end

Timeout::timeout(timeout) do
  Clipboard.notify_change {write}
end


test/test_clipboard.rb
# encoding: utf-8
###########################################################################
# test_clipboard.rb
#
# Test suite for the win32-clipboard library.  This will copy and remove
# data from your clipboard. If your current clipboard data is crucial to
# you, please save it first.
#
# You should run this test case via the 'rake test' task.
###########################################################################
require 'test-unit'
require 'win32/clipboard'
require 'timeout'
include Win32

class TC_Win32_ClipBoard < Test::Unit::TestCase
  test "version is set to expected value" do
    assert_equal('0.6.4', Clipboard::VERSION)
  end

  test "data method basic functionality" do
    assert_respond_to(Clipboard, :data)
    assert_nothing_raised{ Clipboard.data }
    assert_kind_of(String, Clipboard.data)
  end

  test "data method requires proper format" do
    assert_raise(TypeError){ Clipboard.data('test') }
    assert_raise(NameError){ Clipboard.data(CF_FOO) }
  end

  test "get_data is an alias for data" do
    assert_respond_to(Clipboard, :get_data)
    assert_alias_method(Clipboard, :data, :get_data)
  end

  test "set_data basic functionality" do
    assert_respond_to(Clipboard, :set_data)
    assert_nothing_raised{ Clipboard.set_data('foo') }
  end

  test "set_data works with unicode text" do
    assert_nothing_raised{
      Clipboard.set_data('Ηελλας', Clipboard::UNICODETEXT)
    }
  end

  test "set_data requires at least one argument" do
    assert_raise(ArgumentError){ Clipboard.set_data }
  end

  test "set_data requires a valid data format" do
    assert_raise(NameError){ Clipboard.set_data('foo', CF_FOO) }
  end

  test "set and get ascii data as expected" do
    assert_nothing_raised{ Clipboard.set_data('foobar') }
    assert_equal('foobar', Clipboard.data)
  end

  test "set and get unicode data as expected" do
    assert_nothing_raised{
       Clipboard.set_data('Ηελλας', Clipboard::UNICODETEXT)
    }
    assert_equal('Ηελλας', Clipboard.data(Clipboard::UNICODETEXT))
  end

  test "empty method basic functionality" do
    assert_respond_to(Clipboard, :empty)
    assert_nothing_raised{ Clipboard.empty }
  end

  test "clear is an alias for empty" do
    assert_respond_to(Clipboard, :clear)
    assert_alias_method(Clipboard, :clear, :empty)
  end

  test "num_formats basic functionality" do
    assert_respond_to(Clipboard, :num_formats)
    assert_nothing_raised{ Clipboard.num_formats }
    assert_kind_of(Fixnum, Clipboard.num_formats)
  end

  test "num_formats returns an expected value" do
    assert_true(Clipboard.num_formats >= 0)
    assert_true(Clipboard.num_formats < 1000)
  end

  test "num_formats does not accept any arguments" do
    assert_raise(ArgumentError){ Clipboard.num_formats(true) }
  end

  test "register_format basic functionality" do
    assert_respond_to(Clipboard, :register_format)
    assert_nothing_raised{ Clipboard.register_format('Ruby') }
    assert_kind_of(Integer, Clipboard.register_format('Ruby'))
  end

  test "register_format requires a string argument" do
    assert_raises(TypeError){ Clipboard.register_format(1) }
  end

  test "formats method basic functionality" do
    assert_respond_to(Clipboard, :formats)
    assert_nothing_raised{ Clipboard.formats }
    assert_kind_of(Hash, Clipboard.formats)
  end

  # TODO: Why do these fail?
  #test "formats result contains expected values" do
  #  assert_true(Clipboard.formats.size > 0)
  #  assert_true(Clipboard.formats.include?(1))
  #end

  test "formats method does not accept any arguments" do
    assert_raise(ArgumentError){ Clipboard.formats(true) }
  end

  test "format_available basic functionality" do
    assert_respond_to(Clipboard, :format_available?)
    assert_nothing_raised{ Clipboard.format_available?(1) }
  end

  test "format_available returns a boolean value" do
    assert_boolean(Clipboard.format_available?(1))
  end

  test "format_name basic functionality" do
    assert_respond_to(Clipboard, :format_name)
    assert_nothing_raised{ Clipboard.format_name(1) }
  end

  test "format_name returns expected value" do
    format = Clipboard.register_format('HTML Format')
    assert_equal('HTML Format', Clipboard.format_name(format))
    assert_nil(Clipboard.format_name(9999999))
  end

  test "format_name requires a numeric argument" do
    assert_raise(TypeError){ Clipboard.format_name('foo') }
  end

  test "notify_change basic functionality" do
    assert_respond_to(Clipboard, :notify_change)
    def wait(&block)
      Timeout::timeout(5) do
        while @count == 0
          yield if block
          sleep(0.1)
        end
      end
    end

    @count = 0
    t = Thread.start do
      Clipboard.notify_change {@count += 1}
    end
    wait {Clipboard.set_data('ready')}
    @count = 0

    Clipboard.set_data('foo')
    wait
    assert_equal(@count, 1)
    t.kill
  end

  test "expected constants are defined" do
    assert_not_nil(Clipboard::TEXT)
    assert_not_nil(Clipboard::OEMTEXT)
    assert_not_nil(Clipboard::UNICODETEXT)
    assert_not_nil(Clipboard::BITMAP)
    assert_not_nil(Clipboard::DIB)
    assert_not_nil(Clipboard::HDROP)
    assert_not_nil(Clipboard::ENHMETAFILE)
  end
end


lib/win32/html_clipboard.rb
# The Win32 module serves as a namespace only.
#
module Win32

  # The HtmlClipboard class is a subclass of Clipboard that explicitly
  # handles text in HTML format.
  #
  class HtmlClipboard < Clipboard

    private

    # Marker block output
    #--
    # Version: Version of the clipboard.
    #
    # StartHTML: bytecount from the beginning of the clipboard to the
    #    start of the context, or -1 if no context.
    #
    # EndHTML: bytecount from the beginning of the clipboard to the end
    #    of the context, or -1 if no context.
    #
    # StartFragment: bytecount from the beginning of the clipboard to
    #    the start of the fragment.
    #
    # EndFragment: bytecount from the beginning of the clipboard to the
    #    end of the fragment.
    #
    # StartSelection: bytecount from the beginning of the clipboard to
    #    the start of the selection.
    #
    # EndSelection: bytecount from the beginning of the clipboard to the
    #    end of the selection.
    #
    MARKER_BLOCK_OUTPUT =
      "Version:1.0\r\n" \
      "StartHTML:%09d\r\n" \
      "EndHTML:%09d\r\n" \
      "StartFragment:%09d\r\n" \
      "EndFragment:%09d\r\n" \
      "StartSelection:%09d\r\n" \
      "EndSelection:%09d\r\n" \
      "SourceURL:%s\r\n"

    # Extended marker block
    MARKER_BLOCK_EX =
      'Version:(\S+)\s+' \
      'StartHTML:(\d+)\s+' \
      'EndHTML:(\d+)\s+' \
      'StartFragment:(\d+)\s+' \
      'EndFragment:(\d+)\s+' \
      'StartSelection:(\d+)\s+' \
      'EndSelection:(\d+)\s+' \
      'SourceURL:(\S+)'

    # Regular expression for extended marker block
    MARKER_BLOCK_EX_RE = Regexp.new(MARKER_BLOCK_EX, Regexp::MULTILINE) # :nodoc:

    # Standard marker block
    MARKER_BLOCK =
      'Version:(\S+)\s+' \
      'StartHTML:(\d+)\s+' \
      'EndHTML:(\d+)\s+' \
      'StartFragment:(\d+)\s+' \
      'EndFragment:(\d+)\s+' \
      'SourceURL:(\S+)'

    # Regular expression for the standard marker block
    MARKER_BLOCK_RE = Regexp.new(MARKER_BLOCK, Regexp::MULTILINE)

    # Default HTML body
    DEFAULT_HTML_BODY =
      "<!DOCTYPE HTML PUBLIC \"-//W3C//DTD HTML 4.0 Transitional//EN\">\r\n" \
      "<HTML><BODY><!--StartFragment-->%s<!--EndFragment--></BODY></HTML>"

    public

    # Clipboard format value
    CF_HTML = RegisterClipboardFormat("HTML Format")

    def initialize # :nodoc:
      @html      = nil
      @fragment  = nil
      @selection = nil
      @source    = nil
      @version   = nil
    end

    # Returns a boolean indicating whether or not the clipboard contains
    # data in HTML format.
    #
    def self.html_format?
      format_available?(CF_HTML)
    end

    # This method is nearly identical to the Clipboard.data method, but
    # it decodes the data to preserve the HTML formatting.
    #
    def self.data
      begin
        self.open
        if IsClipboardFormatAvailable(CF_HTML)
          handle = GetClipboardData(CF_HTML)
          size   = GlobalSize(handle)
          ptr    = FFI::Pointer.new(:char, handle)

          clip_data = decode_data(ptr.read_bytes(size))
        else
          clip_data = ''
        end
      ensure
        self.close
      end
      clip_data
    end

    # Returns the block marker information for the HTML, or an empty
    # string if there is no clipboard data.
    #
    def self.data_details
      clip_data = data
      string = ""
      unless clip_data.empty?
        string << "prefix=>>>#{@prefix}<<<END"
        string << "version=>>>#{@version}<<<END"
        string << "selection=>>>#{@selection}<<<END"
        string << "fragment=>>>#{@fragment}<<<END"
        string << "html=>>>#{@html}<<<END"
        string << "source=>>>#{@source}<<<END"
      end
      string
    end

    # Put a well-formed fragment of HTML on the clipboard.
    #
    # The +selection+, if provided, must be a literal string within a
    # fragment.
    #
    # The +html+ value, if provided, must be a well formed HTML document
    # that textually contains a fragment and its required markers.
    #
    # The +source+, if provided, should include a scheme (file, http, or
    # https) plus a file name. The default is file:// + __FILE__.
    #
    def self.set_data(fragment, selection=nil, html=nil, source=nil)
      selection ||= fragment
      html      ||= DEFAULT_HTML_BODY % fragment
      source    ||= 'file://' + __FILE__

      fragment_start  = html.index(fragment)
      fragment_end    = fragment_start + fragment.length
      selection_start = html.index(selection)
      selection_end   = selection_start + selection.length

      clip_data = encode_data(
        html,
        fragment_start,
        fragment_end,
        selection_start,
        selection_end,
        source
      )

      self.open
      EmptyClipboard()

      # Global Allocate a movable piece of memory.
      hmem = GlobalAlloc(GHND, clip_data.length + 4)
      mem  = GlobalLock(hmem)
      mem.write_bytes(clip_data, 0, clip_data.size)

      clip_data2 = fragment.gsub(/<[^>]+?>/,'')
      hmem2 = GlobalAlloc(GHND, clip_data2.length + 4)
      mem2  = GlobalLock(hmem2)
      mem2.write_bytes(clip_data2, 0, clip_data2.size)

      # Set the new data
      begin
        if SetClipboardData(CF_HTML, hmem) == 0
          raise SystemCallError.new('SetClipboardData', FFI.errno)
        end

        if SetClipboardData(CF_TEXT, hmem2) == 0
          raise SystemCallError.new('SetClipboardData', FFI.errno)
        end
      ensure
        GlobalFree(hmem)
        GlobalFree(hmem2)
        self.close
      end
      self
    end

    private

    # Encode the data markers into the HTML data
    def self.encode_data(html,frag_start,frag_end,select_start,select_end,src)
      dummy_prefix = MARKER_BLOCK_OUTPUT % [0,0,0,0,0,0,src]
      len_prefix = dummy_prefix.length
      prefix = MARKER_BLOCK_OUTPUT % [len_prefix, html.length + len_prefix,
      frag_start + len_prefix, frag_end + len_prefix,
      select_start + len_prefix, select_end + len_prefix, src]
      prefix + html
    end

    # Decode the given string to figure out the details of the HTML
    # that's on the string.
    #--
    # Try the extended format first, which has an explicit selection.
    # If that fails, try the version without a selection.
    #
    def self.decode_data(src)
      if (matches = MARKER_BLOCK_EX_RE.match(src))
        @prefix = matches[0]
        @version = matches[1]
        @html = src[matches[2].to_i ... matches[3].to_i]
        @fragment = src[matches[4].to_i ... matches[5].to_i]
        @selection = src[matches[6].to_i ... matches[7].to_i]
        @source = matches[8]
      elsif (matches = MARKER_BLOCK_RE.match(src))
        @prefix = matches[0]
        @version = matches[1]
        @html = src[matches[2].to_i ... matches[3].to_i]
        @fragment = src[matches[4].to_i ... matches[5].to_i]
        @source = matches[6]
        @selection = @fragment
      else
        raise 'failed to match block markers'
      end
      @fragment
    end
  end
end


lib/win32/windows/structs.rb
require 'ffi'

module Windows
  module Structs
    extend FFI::Library

    typedef :uchar, :byte
    typedef :ulong, :dword
    typedef :ushort, :word

    class RGBQUAD < FFI::Struct
      layout(
        :rgbBlue, :byte,
        :rgbGreen, :byte,
        :rgbRed, :byte,
        :rgbReserved, :byte,
      )
    end

    class BITMAPINFOHEADER < FFI::Struct
      layout(
        :biSize, :dword,
        :biWidth, :long,
        :biHeight, :long,
        :biPlanes, :word,
        :biBitCount, :word,
        :biCompression, :dword,
        :biSizeImage, :dword,
        :biXPelsPerMeter, :long,
        :biYPelsPerMeter, :long,
        :biClrUsed, :dword,
        :biClrImportant, :dword
      )
    end

    class BITMAPINFO < FFI::Struct
      layout(
        :bmiHeader, BITMAPINFOHEADER,
        :bmiColor, [RGBQUAD, 1]
      )
    end
  end
end


lib/win32/clipboard.rb
require_relative 'windows/constants'
require_relative 'windows/structs'
require_relative 'windows/functions'

# The Win32 module serves as a namespace only.
module Win32

  # The Clipboard class encapsulates functions that relate to the MS Windows clipboard.
  class Clipboard

    include Windows::Constants
    include Windows::Structs
    include Windows::Functions

    extend Windows::Functions
    extend Windows::Structs

    # The version of the win32-clipboard library.
    VERSION = '0.6.4'

    # Clipboard formats

    # Text

    TEXT = 1
    OEMTEXT = 7
    UNICODETEXT = 13

    # Images

    DIB = 8
    BITMAP = 2

    # Metafiles

    ENHMETAFILE = 14

    # Files

    HDROP = 15

    # Empties the contents of the clipboard.
    #
    def self.empty
      begin
        open
        clear_if_already_opened
      ensure
        close
      end
    end

    class << self
      alias clear empty
    end

    # Returns the number of different data formats currently on the clipboard.
    #
    def self.num_formats
      CountClipboardFormats() || 0
    end

    # Returns whether or not +format+ (an integer) is currently available.
    #
    def self.format_available?(format)
      IsClipboardFormatAvailable(format)
    end

    # Returns the data currently in the clipboard. If +format+ is
    # specified, it will attempt to retrieve the data in that format. The
    # default is Clipboard::TEXT.
    #
    # If there is no data in the clipboard, or data is available but the
    # format doesn't match the data, then an empty string is returned.
    #
    # Examples:
    #
    #    # Get some plain text
    #    Win32::Clipboard.data # => e.g. 'hello'
    #
    #    # Get a list of files copied from the Windows Explorer window
    #    Win32::Clipboard.data(Clipboard::HDROP) # => ['foo.rb', 'bar.rb']
    #
    #    # Get a bitmap and write it to another file
    #    File.open('bitmap_copy', 'wb'){ |fh|
    #       fh.write Win32::Clipboard.data(Clipboard::DIB)
    #    }
    #
    def self.data(format = TEXT)
      begin
        open

        if IsClipboardFormatAvailable(format)
          handle = GetClipboardData(format)

          case format
            when UNICODETEXT
              size = GlobalSize(handle)
              ptr  = GlobalLock(handle)
              clip_data = ptr.read_bytes(size).force_encoding('UTF-16LE').encode('UTF-8').strip
            when TEXT, OEMTEXT
              size = GlobalSize(handle)
              ptr  = GlobalLock(handle)

              clip_data = ptr.read_bytes(size).strip

              unless clip_data.ascii_only?
                clip_data.force_encoding('BINARY')
              end
            when HDROP
              clip_data = get_file_list(handle)
            when ENHMETAFILE
              clip_data = get_metafile_data(handle)
            when DIB, BITMAP
              clip_data = get_image_data(handle)
            else
              raise "format '#{format}' not supported"
          end
        else
          clip_data = ''
        end
      ensure
        close
      end

      clip_data
    end

    class << self
      alias get_data data
    end

    # Sets the clipboard contents to the data that you specify. You may
    # optionally specify a clipboard format. The default is Clipboard::TEXT.
    #
    # Example:
    #
    #    # Put the string 'hello' on the clipboard
    #    Win32::Clipboard.set_data('hello')
    #
    #    # Put the image test.bmp on the clipboard
    #    f = File.open("test.bmp", "rb")
    #    str = f.read
    #    Clipboard.set_data(str, Win32::Clipboard::DIB)
    #
    def self.set_data(clip_data, format = TEXT)
      begin
        open
        clear_if_already_opened

        # NULL terminate text or strip header of bitmap file
        case format
          when UNICODETEXT
            clip_data.encode!('UTF-16LE')
            clip_data << "\0".encode('UTF-16LE')
            buf = clip_data
            extra = 4
          when TEXT, OEMTEXT
            clip_data << "\0"
            buf = clip_data
            extra = 4
          when BITMAP, DIB
            buf = clip_data[BITMAP_FILE_HEADER_SIZE..clip_data.length]
            extra = 0
        end

        # Global Allocate a movable piece of memory.
        hmem = GlobalAlloc(GHND, buf.bytesize + extra)
        mptr = GlobalLock(hmem)

        mptr.write_bytes(buf, 0, buf.bytesize)

        # Set the new data
        if SetClipboardData(format, hmem) == 0
          raise SystemCallError.new('SetClipboardData', FFI.errno)
        end
      ensure
        GlobalFree(hmem)
        close
      end
    end

    # Returns a hash of all the current formats, with the format number as
    # the key and the format name as the value for that key.
    #
    # Example:
    #
    #    Win32::Clipboard.formats # => {1 => nil, 49335 => "HTML Format"}
    #
    def self.formats
      formats = {}
      format = 0

      begin
        self.open
        while 0 != (format = EnumClipboardFormats(format))
          buf = FFI::MemoryPointer.new(:char, 80)
          GetClipboardFormatName(format, buf, buf.size)
          string = buf.read_string
          formats[format] = string.empty? ? nil : string
        end
      ensure
        self.close
      end

      formats
    end

    # Returns the corresponding name for the given +format_num+, or nil
    # if it does not exist. You cannot specify any of the predefined
    # clipboard formats (or nil is returned), only registered formats.
    #
    def self.format_name(format_num)
      val = nil
      buf = FFI::MemoryPointer.new(:char, 80)

      begin
        open

        if GetClipboardFormatName(format_num, buf, buf.size) != 0
          val = buf.read_string
        end
      ensure
        close
      end

      val
    end

    # Registers the given +format+ (a String) as a clipboard format, which
    # can then be used as a valid clipboard format. Returns the integer
    # value of the registered format.
    #
    # If a registered format with the specified name already exists, a new
    # format is not registered and the return value identifies the existing
    # format. This enables more than one application to copy and paste data
    # using the same registered clipboard format. Note that the format name
    # comparison is case-insensitive.
    #
    # Registered clipboard formats are identified by values in the range 0xC000
    # through 0xFFFF.
    #
    def self.register_format(format)
      format_value = RegisterClipboardFormat(format)

      if format_value == 0
        raise SystemCallError.new('RegisterClipboardFormat', FFI.errno)
      end

      format_value
    end

    # Sets up a notification loop that will call the provided block whenever
    # there's a change to the clipboard.
    #
    # Example:
    #
    #    Win32::Clipboard.notify_change{
    #       puts "There's been a change in the clipboard contents!"
    #    }
    #--
    # We skip the first notification because the very act of attaching the
    # new window causes it to trigger once.
    #
    def self.notify_change(&block)
      name   = 'ruby-clipboard-' + Time.now.to_s
      handle = CreateWindowEx(0, 'static', name, 0, 0, 0, 0, 0, 0, 0, 0, nil)

      next_viewer = SetClipboardViewer(handle)

      if next_viewer.nil?
        raise SystemCallError.new('SetClipboardViewer', FFI.errno)
      end

      SetWindowLongPtr(handle, GWL_USERDATA, next_viewer)

      ObjectSpace.define_finalizer(self, proc{self.close_window(handle)})

      @first_notify = true

      wnd_proc = FFI::Function.new(:uintptr_t, [:uintptr_t, :uint, :uintptr_t, :uintptr_t]) do |hwnd, umsg, wparam, lparam|
        case umsg
           when WM_DESTROY
             next_viewer = GetWindowLongPtr(hwnd, GWL_USERDATA)
             ChangeClipboardChain(hwnd, next_viewer)
             PostQuitMessage(0)
             rv = 0
           when WM_DRAWCLIPBOARD
             yield unless @first_notify
             next_viewer = GetWindowLongPtr(hwnd, GWL_USERDATA)
             if next_viewer != 0
               PostMessage(next_viewer, umsg, wparam, lparam)
             end
             rv = 0
           when WM_CHANGECBCHAIN
             next_viewer = GetWindowLongPtr(hwnd, GWL_USERDATA)
             next_viewer = lparam if next_viewer == wparam
             SetWindowLongPtr(hwnd, GWL_USERDATA, next_viewer)
             if next_viewer != 0
               PostMessage(next_viewer, umsg, wparam, lparam)
             end
             rv = 0
           else
             rv = DefWindowProc(hwnd, umsg, wparam, lparam)
        end

        @first_notify = false

        rv
      end

      if SetWindowLongPtr(handle, GWL_WNDPROC, wnd_proc.address) == 0
        raise SystemCallError.new('SetWindowLongPtr', FFI.errno)
      end

      msg = FFI::MemoryPointer.new(:char, 100)

      while true
        while PeekMessage(msg, handle, 0, 0, 1)
          TranslateMessage(msg)
          DispatchMessage(msg)
        end
        sleep 0.1
      end
    end

    private

    # sizeof(BITMAPFILEHEADER)

    BITMAP_FILE_HEADER_SIZE = 14

    # sizeof(BITMAPINFOHEADER)

    BITMAP_INFO_HEADER_SIZE = 40

    # Opens the clipboard and prevents other applications from modifying
    # the clipboard content until it is closed.
    #
    def self.open
      unless OpenClipboard(0)
        raise SystemCallError.new('OpenClipboard', FFI.errno)
      end
    end

    # Close the clipboard
    #
    def self.close
      unless CloseClipboard()
        raise SystemCallError.new('CloseClipboard', FFI.errno)
      end
    end

    # Clear the clipboard that is already opened
    #
    def self.clear_if_already_opened
      unless EmptyClipboard()
        raise SystemCallError.new('EmptyClipboard', FFI.errno)
      end
    end

    # Get data for enhanced metadata files
    #
    def self.get_metafile_data(handle)
      buf_size = GetEnhMetaFileBits(handle, 0, nil)
      buf = FFI::MemoryPointer.new(:char, buf_size)
      GetEnhMetaFileBits(handle, buf.size, buf)
      buf.read_string
    end

    # Get data for bitmap files
    #
    def self.get_image_data(handle)
      buf = nil

      begin
        ptr = GlobalLock(handle)
        buf_size = GlobalSize(handle)

        bmi = BITMAPINFO.new(ptr)

        file_size = buf_size + BITMAP_FILE_HEADER_SIZE

        # Calculate the header size
        case bmi[:bmiHeader][:biBitCount]
          when 1
            table_size = 2
          when 4
            table_size = 16
          when 8
            table_size = 256
          when 16, 32
            if bmi[:bmiHeader][:biCompression] == BI_RGB
              table_size = bmi[:bmiHeader][:biClrUsed]
            elsif bmi[:bmiHeader][:biCompression] == BI_BITFIELDS
              table_size = bmi[:bmiHeader][:biClrUsed]
              if bmi[:bmiHeader][:biSize] == BITMAP_INFO_HEADER_SIZE
                # Add bit fields size
                table_size += 3
              end
            else
              raise "invalid bit/compression combination"
            end
          when 24
            table_size = bmi[:bmiHeader][:biClrUsed]
          else
            raise "invalid bit count"
        end # case

        # offset = sizeof(BITMAPFILEHEADER)
        #          + sizeof(BITMAPINFOHEADER or BITMAPV4HEADER or BITMAPV5HEADER)
        #          + (color table size)
        offset = BITMAP_FILE_HEADER_SIZE + bmi[:bmiHeader][:biSize] + (table_size * 4)

        buf = "\x42\x4D" + [file_size].pack('L') + 0.chr * 4 + [offset].pack('L') + ptr.read_bytes(buf_size)
      ensure
        GlobalUnlock(handle) if handle
      end

      buf
    end

    # Get and return an array of file names that have been copied.
    #
    def self.get_file_list(handle)
      array = []
      count = DragQueryFileA(handle, 0xFFFFFFFF, nil, 0)

      0.upto(count - 1){ |i|
        length = DragQueryFileA(handle, i, nil, 0) + 1
        buf = FFI::MemoryPointer.new(:char, length)
        DragQueryFileA(handle, i, buf, buf.size)
        array << buf.read_string
      }

      array
    end

    # Close a window.(for self.notify_change)
    #
    def self.close_window(handle)
      SendMessage(handle, WM_CLOSE, 0, 0)
    end
  end
end

require File.join(File.dirname(__FILE__), 'html_clipboard')
